File: 0684_Redundant Connection/RedundantConnectionDFS.java
class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        int n = edges.length;
        List<List<Integer>> graph = new ArrayList<>();
        for(int i=0;i<=n;i++) graph.add(new ArrayList<Integer>());
        
        for(int[] edge : edges) {
            int u = edge[0], v = edge[1];
            if(DFS(graph, u, v, new HashSet<Integer>())) return edge;
            graph.get(u).add(v);
            graph.get(v).add(u);
        }
        
        return new int[2];
    }
    
    private boolean DFS(List<List<Integer>> graph, int u, int v, Set<Integer> visited) {
        if(u == v) return true;
        visited.add(u);
        for(Integer next : graph.get(u)) {
            if(visited.contains(next)) continue;
            if(DFS(graph, next, v, visited)) return true;
        }
        
        return false;
    }
}

File: 0684_Redundant Connection/RedundantConnectionDFS.cs
public class Solution {
    public int[] FindRedundantConnection(int[][] edges) {
        var adjacencyList = new List<List<int>>();
        
        for(int i=0;i<edges.Length+1;i++)
        {
            adjacencyList.Add(new List<int>());
        }
        
        for(int i=0;i<edges.Length;i++)
        {
            var u = edges[i][0];
            var v = edges[i][1];
            
            if(DFS(adjacencyList, u, v, new HashSet<int>())) return edges[i];
            
            adjacencyList[u].Add(v);
            adjacencyList[v].Add(u);
        }
        
        return new int[]{};
    }
    
    private bool DFS(List<List<int>> adjacencyList, int u, int v, HashSet<int> visited)
    {
        if(u == v) return true;
        visited.Add(u);
        
        foreach(var next in adjacencyList[u])
        {
            if(visited.Contains(next)) continue;
            if(DFS(adjacencyList, next, v, visited)) return true;
        }
        
        return false;
    }
}